### PR TITLE
task fuzzy matching and verify_cot

### PR DESCRIPTION
**Pull request by escapewindow**
This is a two-part PR, neatly separated into two discrete commits.

1. enable task fuzzy-matching: to support task retriggers, we can now match a task against the task-graph definitions without requiring the taskId be present in the original task graph.  We would still need a tool to be able to retrigger large portions of the task graph with new task ids if desired, but this should support one-offs as is.
2. We had a helper script named `scriptworker/test/data/verify_cot.py` that helped me test the chain of trust verification without having to run a scriptworker instance.  This is now an entry point for easier use.

### CODE REVIEW

**Comment by escapewindow** [comment]
@lundjordan , @JohanLorenzo I know you're both in Paris working on shipit.  I'll be on PTO next week (Nov21).  I'm currently thinking I'll leave this for non-urgent review, hoping to merge by the time I'm back on Nov28.

**Comment by escapewindow** [comment]
Rebased due to docs landing.

File: scriptworker/cot/verify.py [hunk 1]
```diff
@@ -678,7 +687,46 @@ def verify_link_in_task_graph(chain, decision_link, task_link):
...
+
+
+# verify_link_in_task_graph {{{1
+def verify_link_in_task_graph(chain, decision_link, task_link):
+    """Compare the runtime task definition against the decision task graph.
+
+    If the ``task_link.task_id`` is in the task graph, match directly against
+    that task definition.  Otherwise, "fuzzy match" by trying to match against
+    any definition in the task graph.  This is to support retriggers, where
+    the task definition stays the same, but the datestrings and taskIds change.
```
Comment (JohanLorenzo) [inline scriptworker/cot/verify.py]: Nice piece of doc!

File: scriptworker/test/test_cot_verify.py [hunk 2]
```diff
@@ -598,6 +600,19 @@ def test_verify_link_in_task_graph(chain, decision_link, build_link):
     cotverify.verify_link_in_task_graph(chain, decision_link, build_link)
 
 
+def test_verify_link_in_task_graph_fuzzy_match(chain, decision_link, build_link):
+    chain.links = [decision_link, build_link]
+    decision_link.task_graph = {
+        'bogus-task-id': {
```
Comment (JohanLorenzo) [inline scriptworker/test/test_cot_verify.py]: TIL: "bogus" doesn't refer to "bug" #EnglishLessonOfTheDay

File: scriptworker/cot/verify.py [hunk 3]
```diff
@@ -678,7 +687,46 @@ def verify_link_in_task_graph(chain, decision_link, task_link):
...
+        CoTError: on failure.
+    """
+    log.info("Verifying the {} {} task definition is part of the {} {} task graph...".format(
+        task_link.name, task_link.task_id, decision_link.name, decision_link.task_id
+    ))
+    if task_link.task_id in decision_link.task_graph:
+        graph_defn = deepcopy(decision_link.task_graph[task_link.task_id])
+        verify_task_in_task_graph(task_link, graph_defn)
+        return
+    # Fall back to fuzzy matching to support retriggers: the taskId and
```
Comment (JohanLorenzo) [inline scriptworker/cot/verify.py]: Nit: Maybe add an `info` or a `warning` to tell users that we go the fuzzy way.

File: scriptworker/cot/verify.py [hunk 4]
```diff
@@ -1076,3 +1124,61 @@ def callback(match):
...
+            context.config = dict(deepcopy(DEFAULT_CONFIG))
+            context.config.update({
+                'artifact_dir': os.path.join(tmp, 'artifacts'),
+                'base_gpg_home_dir': os.path.join(tmp, 'gpg'),
+                'verify_cot_signature': False,
+            })
+            cot = ChainOfTrust(context, 'signing', task_id=opts.task_id)
+            loop.run_until_complete(verify_chain_of_trust(cot))
+            pprint.pprint(cot.dependent_task_ids())
+            print("Cot task_id: {}".format(cot.task_id))
```
Comment (JohanLorenzo) [inline scriptworker/cot/verify.py]: Nit: `log` is not used

File: scriptworker/cot/verify.py [hunk 5]
```diff
@@ -1076,3 +1124,61 @@ def callback(match):
...
+            for link in cot.links:
+                print("task_id: {}".format(link.task_id))
```
Comment (JohanLorenzo) [inline scriptworker/cot/verify.py]: Nit: Same comment about `log`

File: scriptworker/cot/verify.py (hunk 3 above)
Comment (escapewindow) [inline scriptworker/cot/verify.py]: Done.

File: scriptworker/cot/verify.py (hunk 4 above)
Comment (escapewindow) [inline scriptworker/cot/verify.py]: Fixed.  This was for a standalone commandline tool, so it's not as bad, but it's easy enough to use `log`.

**Comment by escapewindow** [comment]
The `USE_SCCACHE` env var is new and breaking current chain of trust verification: https://public-artifacts.taskcluster.net/UT7a4_UEQbSFk8VOQ7BNTw/0/public/cot/audit.log

**Pull request merged**